pandas/core/indexes/base.py

```diff
@@ -2263,19 +2263,15 @@ def __or__(self, other):
     def __xor__(self, other):
         return self.symmetric_difference(other)
 
-    def _get_consensus_name(self, other):
+    def _get_setop_name_object(self, other):
         """
-        Given 2 indexes, give a consensus name meaning
-        we take the not None one, or None if the names differ.
-        Return a new object if we are resetting the name
+        Given 2 indexes, give a setop name and object, meaning
```

Comment:
instead why can't we just use _wrap_setop_result? instead of having 2 nearly identical funtions

Reply:
@jreback I think this is needed (and I'll rename it and document it accordingly) in the case you are computing `i1.union(i2)`, where `i1.equals(i2)==True`.  In this case, you don't want to use `_shallow_copy()` because you want to return `i1`, unless the names of `i1` and `i2` are different. There are some tests for joins where if you are doing a join and the index is the same, then you can just return the index, rather than returning a brand new index, which is what `_shallow_copy()` would do.

I'll also check the other `Index` subclasses to see if they are doing this right.  For example, for `RangeIndex`, it is not handling the names correctly.

Reply:
hmm, I think we do this elsewhere as well (e.g. when you reindex to yourself). can you add docmentation about this to the setops themselves. @TomAugspurger @jorisvandenbossche what do you think about this?

Reply:
I don't have a strong preference, but it'd be good to settle on consistent behavior between union and intersection (I think). What's the proposal here?

1. named & same name -> name
2. named & other name -> None?
3. named & unnamed -> name?
4. unnamed & named -> name?
5. unamed & unnamed -> None

Reply:
My bad.

I don't have a strong preference, but I would tend towards always shallow copying? I hope people aren't relying on `(idx | idx) is idx` being true.

Reply:
@TomAugspurger With respect to your naming, that is *not* what will happen.  In the cases of named and unnamed, the result will be `None`.  That's because if you do
```
j1 = pd.Index([1,2], name='j1')
j2 = pd.Index([], name='j2')
j3 = pd.Index([], name='j3')
```
then `j1.union(j2).union(j3) returns Int64Index([1, 2], dtype='int64', name='j3')`, while just changing the order to `j3.union(j1).union(j2)` returns` Int64Index([1, 2], dtype='int64', name='j2')`. So the behavior will be to only preserve the names if they are equal, and the result of these two union examples will have `None` as the name.

@jreback I can change to shallow copying, but I'll have to modify some other tests that are depending on the result not being a shallow copy.

Reply:
yeah I think its better to _shallow_copy these, yes I understand some tests may change because of this (IOW they will no longer be ``is``, just ``.equals``)